src/nexusrpc/handler/_common.py

```diff
@@ -0,0 +1,175 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import (
+    Generic,
+    Mapping,
+    Optional,
+    Sequence,
+)
+
+from nexusrpc import Link
+from nexusrpc.types import OutputT
+
+
+class HandlerErrorType(Enum):
+    """Nexus handler error types.
+
+    See https://github.com/nexus-rpc/api/blob/main/SPEC.md#predefined-handler-errors
+    """
+
+    BAD_REQUEST = "BAD_REQUEST"
+    UNAUTHENTICATED = "UNAUTHENTICATED"
+    UNAUTHORIZED = "UNAUTHORIZED"
+    NOT_FOUND = "NOT_FOUND"
+    RESOURCE_EXHAUSTED = "RESOURCE_EXHAUSTED"
+    INTERNAL = "INTERNAL"
+    NOT_IMPLEMENTED = "NOT_IMPLEMENTED"
+    UNAVAILABLE = "UNAVAILABLE"
+    UPSTREAM_TIMEOUT = "UPSTREAM_TIMEOUT"
+
+
+class HandlerError(Exception):
+    """
+    A Nexus handler error.
+
+    This exception is used to represent errors that occur during the handling of a
+    Nexus operation that should be reported to the caller as a handler error.
+    """
+
+    def __init__(
+        self,
+        message: str,
+        *,
+        type: HandlerErrorType,
+        cause: Optional[BaseException] = None,
+        # Whether this error should be considered retryable. If not specified, retry
+        # behavior is determined from the error type. For example, INTERNAL is retryable
+        # by default unless specified otherwise.
+        retryable: Optional[bool] = None,
+    ):
+        """
+        Initializes a new HandlerError.
+
+        :param message: A descriptive message for the error. This will become the `message`
+                        in the resulting Nexus Failure object.
+        :param type: The type of handler error.
+        :param cause: The original exception that caused this handler error, if any.
+                      This will be encoded in the `details` of the Nexus Failure object.
+        :param retryable: Whether this error should be retried. If not
+                          provided, the default behavior for the error type is used.
+        """
+        super().__init__(message)
+        self.__cause__ = cause
```

Comment:
Hrmm, Python exceptions do not usually accept cause in constructors, they usually rely on the `from` clause during `raise` or in the rare case of needing to create the exception with a cause without a `raise`, users manually set `__cause__` themselves after creation. Having it in constructor is confusing, especially in cases where it's unclear it may override.

Reply:
Yes I was taking the proto schema too literally here. Removed `cause` from constructor; `raise from` can be used everywhere where `cause` was being used.